# Attack 4 - Endpoints Flood: Comparison

Loads baseline and attack4 CSV exports and compares them (volume, rates, deltas).
See `attacks/attack4-endpoints-flood.md` for the attack description and deployment manifests.

- Baseline capture: `exports/baseline/`
- Attack capture: `exports/attack4-endpoints-flooding/`
- Comparison logic: `compare_logs.py`

In [1]:
from pathlib import Path
from IPython.display import display

ATTACK_DIR = Path("../exports/attack4-endpoints-flooding")
ATTACK_SUFFIX = "attack4"

%run -i compare_logs.py

print("=== Resume ===")
display(summary)

print("\n=== Events per second (stats) ===")
display(eps_stats)

print("\n=== Resource x Verb ===")
display(resource_x_verb)

for field, df in results.items():
    print(f"\n=== {field} ===")
    display(df)

        métrique  baseline  attack  variation
0   total_events      2089   16928      14839
1  patch_on_pods         0    2068       2068
=== Resume ===


,métrique,baseline,attack,variation
0,total_events,2089,16928,14839
1,patch_on_pods,0,2068,2068



=== Events per second (stats) ===


,métrique,baseline,attack,variation
0,max_events_par_sec,25.0,177.00,152.00
1,moyenne_events_par_sec,7.6,61.88,54.28



=== Resource x Verb ===


,resource,verb,count_baseline,count_attack,variation
0,deployments,update,0,3129,3129
1,replicasets,update,0,3119,3119
2,pods,patch,0,2068,2068
3,pods,get,0,2063,2063
4,endpoints,update,0,2032,2032
...,...,...,...,...,...
71,services,watch,24,9,-15
72,clusterinformations,get,60,40,-20
73,endpointslices,get,58,38,-20
74,endpoints,get,58,38,-20



=== verb ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,update,710,10697,142.0,2139.4,9987,1997.4
1,patch,3,3169,0.6,633.8,3166,633.2
2,get,1063,2806,212.6,561.2,1743,348.6
3,watch,309,214,61.8,42.8,-95,-19.0
4,create,0,18,0.0,3.6,18,3.6
5,list,4,12,0.8,2.4,8,1.6



=== objectRef_resource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,pods,12,4167,2.4,833.4,4155,831.0
1,deployments,3,3132,0.6,626.4,3129,625.8
2,replicasets,6,3122,1.2,624.4,3116,623.2
3,endpoints,64,2070,12.8,414.0,2006,401.2
4,endpointslices,70,2012,14.0,402.4,1942,388.4
...,...,...,...,...,...,...,...
56,ipreservations,3,0,0.6,0.0,-3,-0.6
57,rolebindings,3,0,0.6,0.0,-3,-0.6
58,priorityclasses,6,0,1.2,0.0,-6,-1.2
59,prioritylevelconfigurations,3,0,0.6,0.0,-3,-0.6



=== objectRef_namespace ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,attacker-ns,0,15526,0.0,3105.2,15526,3105.2
1,kube-system,643,412,128.6,82.4,-231,-46.2
2,default,116,75,23.2,15.0,-41,-8.2
3,kube-node-lease,88,53,17.6,10.6,-35,-7.0
4,safe-ns,0,3,0.0,0.6,3,0.6



=== objectRef_subresource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,status,3,8318,0.6,1663.6,8315,1663.0



=== responseStatus_code ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,200,990,2867,198.0,573.4,1877,375.4



=== authorization_decision ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,allow,990,2867,198.0,573.4,1877,375.4



=== authorization_reason ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,"RBAC: allowed by ClusterRoleBinding ""system:co...",0,1091,0.0,218.2,1091,218.2
1,"RBAC: allowed by ClusterRoleBinding ""system:co...",0,1067,0.0,213.4,1067,213.4
2,"RBAC: allowed by ClusterRoleBinding ""system:pu...",330,212,66.0,42.4,-118,-23.6
3,"RBAC: allowed by ClusterRoleBinding ""system:ku...",236,150,47.2,30.0,-86,-17.2
4,"RBAC: allowed by ClusterRoleBinding ""system:ku...",161,110,32.2,22.0,-51,-10.2
5,"RBAC: allowed by ClusterRoleBinding ""system:di...",65,47,13.0,9.4,-18,-3.6
6,"RBAC: allowed by ClusterRoleBinding ""calico-ku...",30,20,6.0,4.0,-10,-2.0
7,"RBAC: allowed by ClusterRoleBinding ""kubeadm:c...",0,15,0.0,3.0,15,3.0
8,"RBAC: allowed by ClusterRoleBinding ""kubeadm:n...",14,5,2.8,1.0,-9,-1.8
9,"RBAC: allowed by ClusterRoleBinding ""system:vo...",5,2,1.0,0.4,-3,-0.6



=== sourceIPs ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,10.10.10.10,1572,11307,314.4,2261.4,9735,1947.0
1,10.10.10.11,83,5321,16.6,1064.2,5238,1047.6
2,::1,210,140,42.0,28.0,-70,-14.0
3,192.168.62.169,0,79,0.0,15.8,79,15.8
4,10.10.10.12,89,66,17.8,13.2,-23,-4.6
5,192.168.62.170,0,3,0.0,0.6,3,0.6
6,10.0.2.15,6,0,1.2,0.0,-6,-1.2
7,192.168.62.160,9,0,1.8,0.0,-9,-1.8
8,192.168.62.161,120,0,24.0,0.0,-120,-24.0



=== user_groups ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:authenticated,1429,16493,285.8,3298.6,15064,3012.8
1,system:serviceaccounts:kube-system,251,10386,50.2,2077.2,10135,2027.0
2,system:serviceaccounts,251,10386,50.2,2077.2,10135,2027.0
3,system:nodes,217,5445,43.4,1089.0,5228,1045.6
4,system:unauthenticated,660,423,132.0,84.6,-237,-47.4
5,system:masters,210,140,42.0,28.0,-70,-14.0
6,kubeadm:cluster-admins,0,31,0.0,6.2,31,6.2



=== user_username ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:node:k8s-w1,71,5318,14.2,1063.6,5247,1049.4
1,system:serviceaccount:kube-system:deployment-c...,0,3129,0.0,625.8,3129,625.8
2,system:serviceaccount:kube-system:replicaset-c...,0,3119,0.0,623.8,3119,623.8
3,system:serviceaccount:kube-system:endpoint-con...,0,2032,0.0,406.4,2032,406.4
4,system:serviceaccount:kube-system:endpointslic...,0,1964,0.0,392.8,1964,392.8
5,system:anonymous,660,423,132.0,84.6,-237,-47.4
6,system:kube-controller-manager,428,272,85.6,54.4,-156,-31.2
7,system:kube-scheduler,323,219,64.6,43.8,-104,-20.8
8,system:apiserver,210,140,42.0,28.0,-70,-14.0
9,system:serviceaccount:kube-system:calico-kube-...,120,79,24.0,15.8,-41,-8.2



=== userAgent ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kubelet/v1.30.14 (linux/amd64) kubernetes/9e18483,217,5445,43.4,1089.0,5228,1045.6
1,kube-controller-manager/v1.30.14 (linux/amd64)...,0,3129,0.0,625.8,3129,625.8
2,kube-controller-manager/v1.30.14 (linux/amd64)...,0,3119,0.0,623.8,3119,623.8
3,kube-controller-manager/v1.30.14 (linux/amd64)...,0,2032,0.0,406.4,2032,406.4
4,kube-controller-manager/v1.30.14 (linux/amd64)...,0,1964,0.0,392.8,1964,392.8
5,kube-probe/1.30,660,423,132.0,84.6,-237,-47.4
6,kube-scheduler/v1.30.14 (linux/amd64) kubernet...,296,195,59.2,39.0,-101,-20.2
7,kube-controller-manager/v1.30.14 (linux/amd64)...,296,186,59.2,37.2,-110,-22.0
8,kube-apiserver/v1.30.14 (linux/amd64) kubernet...,210,140,42.0,28.0,-70,-14.0
9,kube-controller-manager/v1.30.14 (linux/amd64)...,96,54,19.2,10.8,-42,-8.4


## Results summary

| Metric | Baseline | Attack | Variation | Factor |
|---|---|---|---|---|
| Total audit events | 2089 | 16928 | +14839 | x8.1 |
| Events/sec (average) | 7.6 | 61.9 | +54.3 | x8.1 |
| Events/sec (peak) | 25 | 177 | +152 | x7.1 |
| `objectRef.subresource: status` | 3 | 8318 | +8315 | x2773 |
| `authorization.decision: allow` | 990 | 2867 | +1877 | x2.9 |

| Controller account (`userAgent`) | Baseline | Attack |
|---|---|---|
| deployment-controller | 0 | 3129 |
| replicaset-controller | 0 | 3119 |
| endpoint-controller | 0 | 2032 |
| endpointslice-controller | 0 | 1964 |

| Namespace | Baseline | Attack |
|---|---|---|
| attacker-ns | 0 | 15526 |
| safe-ns | 0 | 3 |

---

## Analysis of results

The breakdown by `(resource, verb)` reveals two groups of affected objects that correspond precisely to the two data dependencies exploited by the attack (sections 4.4 and 4.5). On one side, `replicasets` (+3119) and `deployments` (+3129) materialize the pod-replicaset-deployment dependency (Fig. 6): every readiness flip on a pod forces the `kube-apiserver` to propagate the change to the parent `ReplicaSet` and then `Deployment`. On the other side, `endpoints` (+2032) and `endpointslices` (+1964) materialize the pod-endpoint-service dependency (Fig. 7), specific to this attack. This observation confirms empirically, with figures to back it up, that the compromise of a single resource (the pod) amplifies its impact across the entire graph of dependent objects, since both dependency chains show up in our logs with volumes of the same order of magnitude (roughly 2000 to 3100 events each), consistent with the fact that they are triggered by the same source event (the readiness flip) and propagate in parallel rather than sequentially.

The shift of `objectRef.subresource: status` from 0.1% to 49% of the total volume is the clearest signature of this reconciliation mechanism: nearly all the noise generated corresponds to state updates, not structural changes to the cluster, which is consistent with an attack that creates or deletes no business object after the initial deployment phase, but instead shakes the `status` field across the entire dependency graph.

The simultaneous appearance of the four controller accounts (`deployment-controller`, `replicaset-controller`, `endpoint-controller`, `endpointslice-controller`), each entirely silent in the baseline, is a direct and unambiguous fingerprint of this dual propagation: their combined volume (10244 events) accounts for 60% of the total generated by the attack on its own, making them the main contributor to the observed signal, more so than the attacker's own requests.

Finally, the fact that `authorization.decision` stays at 100% `allow` throughout the attack confirms that it exploits no authorization flaw: it stays, start to finish, within the default RBAC permissions of a pod's `service account token` (Table 1 of the paper: `pods/status: patch`, `services`/`endpoints: watch`), consistent with the paper's threat model (section 3.1) where the attacker needs no elevated privileges.

---

## Detectability analysis

**1. Appearance of service accounts absent from the baseline.** A system account that is usually silent and suddenly becomes active is an indicator with a very low false positive rate: binary detection by presence/absence, not a fragile statistical threshold.

**2. Concentration on `objectRef.subresource: status`.** A rule such as "more than X% of events in an interval target the `status` subresource" would be a good candidate for a Wazuh rule, cheap to evaluate.

**3. Volume and rate.** The x8 factor on `total_events` and `events/sec` largely exceeds the background noise measured in the baseline (peaks of 25/sec or less in the baseline versus 177/sec during the attack). A simple threshold on the event rate would detect this attack unambiguously.

### Detectability limitations to note

- **RBAC decision always `allow`**: a rule based solely on authorization denials would be blind to this attack. The detector must rely on *volume*, not on the *legitimacy* of the calls.
- **Namespace confinement**: if the attacker spreads activity across several namespaces or inside a shared legitimate namespace, the `objectRef.namespace` indicator loses its discriminating power. In that case, detection would have to rely solely on indicators 1 and 2, which remain valid regardless of namespace.
- **Dependency on the temporal observation point**: as observed during capture (the gap between `@timestamp` and `requestReceivedTimestamp` under load), a real-time detector based on indexing time can temporarily underestimate the real event rate during load spikes.